runtime/src/storage/iouring.rs

```diff
@@ -0,0 +1,477 @@
+        Poll::Pending
+    }
+}
+
+/// Background task that polls for completed work and notifies waiters on completion.
+/// The user data field of all operations received on `receiver` will be ignored.
+async fn do_work(mut receiver: mpsc::Receiver<(SqueueEntry, oneshot::Sender<i32>)>) {
+    let mut ring = IoUring::new(IOURING_SIZE).expect("failed to create io_uring");
+    let mut next_work_id: u64 = 0;
+    // Maps a work ID to the sender that we will send the result to.
+    let mut waiters: HashMap<_, oneshot::Sender<i32>> =
+        HashMap::with_capacity(IOURING_SIZE as usize);
+
+    loop {
+        let completed_work_fut = NextCompletionFuture::new(&mut ring);
+        let new_work_fut = if waiters.len() < IOURING_SIZE as usize {
+            Either::Left(receiver.next())
+        } else {
+            // We're at the limit for maximum number of ongoing operations.
+            // Wait for a completion to free up space.
+            Either::Right(futures::future::pending())
+        };
+
+        select! {
+            next_work = new_work_fut => {
+                let Some((mut op,sender)) = next_work else {
+                    // Channel closed, exit the loop
+                    break;
+                };
+
+                // Assign a unique id
+                let work_id = next_work_id;
+                op = op.user_data(work_id);
+                next_work_id = next_work_id.wrapping_add(1);
+
+                // We'll send the result of this operation to `sender`.
+                waiters.insert(work_id, sender);
+
+                // Submit the operation to the ring
+                unsafe {
+                    ring.submission().push(&op).expect("unable to push to queue");
+                }
+            },
+            completed_work = completed_work_fut => {
+                let work_id = completed_work.user_data();
+                let result = completed_work.result();
+                let sender = waiters.remove(&work_id).expect("work is missing");
+                // Notify with the result of this operation
+                let _ =  sender.send(result);
+            },
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct Storage {
+    storage_directory: PathBuf,
+    io_sender: mpsc::Sender<(SqueueEntry, oneshot::Sender<i32>)>,
+}
+
+impl Storage {
+    /// Returns a new `Storage` instance.
+    /// The `Spawner` is used to spawn the background task that handles IO operations.
+    pub fn start<S: Spawner>(cfg: &Config, spawner: S) -> Self {
+        let (io_sender, receiver) =
+            mpsc::channel::<(SqueueEntry, oneshot::Sender<i32>)>(IOURING_SIZE as usize);
+
+        let storage = Storage {
+            storage_directory: cfg.storage_directory.clone(),
+            io_sender,
+        };
+        spawner.spawn(|_| do_work(receiver));
+        storage
+    }
+}
+
+impl crate::Storage for Storage {
+    type Blob = Blob;
+
+    async fn open(&self, partition: &str, name: &[u8]) -> Result<Blob, Error> {
```

Comment:
I wonder if we can turn these `fs` calls into utils that can be shared between tokio storage impls